src/helpers/task-update.ts

```diff
@@ -1,45 +1,90 @@
 import { getTaskAssignmentDetails } from "./task-metadata";
 
-export async function updateTaskReminder(context: Context, repo: ListForOrg["data"][0], issue: ListIssueForRepo) {
-  const { logger } = context;
-
-  let deadlineWithThreshold, reminderWithThreshold;
-  const now = DateTime.now();
+const getMostRecentActivityDate = (assignedEventDate: DateTime, activityEventDate?: DateTime): DateTime => {
+  return activityEventDate && activityEventDate > assignedEventDate ? activityEventDate : assignedEventDate;
+};
 
+export async function updateTaskReminder(context: ContextPlugin, repo: ListForOrg["data"][0], issue: ListIssueForRepo) {
+  const {
+    octokit,
+    logger,
+    config: { eventWhitelist, warning, disqualification },
+  } = context;
   const handledMetadata = await getTaskAssignmentDetails(context, repo, issue);
+  const now = DateTime.local();
 
-  if (handledMetadata) {
-    const handledDeadline = await getDeadlineWithThreshold(context, handledMetadata, issue);
-    if (handledDeadline) {
-      deadlineWithThreshold = handledDeadline.deadlineWithThreshold;
-      reminderWithThreshold = handledDeadline.reminderWithThreshold;
-
-      logger.info(`Handling metadata and deadline for ${issue.html_url}`, {
-        initialDeadline: DateTime.fromISO(handledMetadata.startPlusLabelDuration).toLocaleString(DateTime.DATETIME_MED),
-        now: now.toLocaleString(DateTime.DATETIME_MED),
-        reminderWithThreshold: reminderWithThreshold.toLocaleString(DateTime.DATETIME_MED),
-        deadlineWithThreshold: deadlineWithThreshold.toLocaleString(DateTime.DATETIME_MED),
-      });
-    }
-  }
+  if (!handledMetadata) return;
+
+  const assignmentEvents = await octokit.paginate(octokit.rest.issues.listEvents, {
+    owner: repo.owner.login,
+    repo: repo.name,
+    issue_number: issue.number,
+  });
 
-  if (!deadlineWithThreshold || !reminderWithThreshold) {
-    logger.error(`Failed to handle metadata or deadline for ${issue.html_url}`);
-    return false;
+  const assignedEvent = assignmentEvents
+    .filter((o) => o.event === "assigned" && handledMetadata.taskAssignees.includes(o.actor.id))
+    .sort((a, b) => DateTime.fromISO(b.created_at).toMillis() - DateTime.fromISO(a.created_at).toMillis())
+    .shift();
+
+  if (!assignedEvent) {
+    logger.error(`Failed to update activity for ${issue.html_url}, there is no assigned event.`);
+    return;
   }
 
-  if (now >= deadlineWithThreshold) {
-    // if the issue is past due, we should unassign the user
-    await unassignUserFromIssue(context, issue);
-  } else if (now >= reminderWithThreshold) {
-    // if the issue is within the reminder threshold, we should remind the assignees
-    await remindAssigneesForIssue(context, issue);
+  const activityEvent = (await getAssigneesActivityForIssue(context, issue, handledMetadata.taskAssignees))
+    .filter((o) => eventWhitelist.includes(o.event as TimelineEvent))
+    .shift();
```

Comment:
This looks like this can shift an `undefined` value, I don't think this is type safe. Check if its undefined and throw a specific error for that if you must.

Reply:
`undefined` is checked, the shift is on purpose, to check if there was any activity, and if not define the last activity as when the user got assigned.